### PR TITLE
Fix copy bug

### LINKED ISSUE #2232: Long string & lists cause SIGSEGV Heap Buffer Overflow
When importing long lists or strings from CSV, we should hit an exception and return a useful exception to the user. Instead, we get a heap-based buffer overflow and segfault. Test files attached. To use `long-string.csv`, run:

```
CREATE NODE TABLE test(A STRING, PRIMARY KEY(A));
COPY test FROM "long-string.csv";
```

Building with ASAN recommended.
[long-list.csv](https://github.com/kuzudb/kuzu/files/13030483/long-list.csv)
[long-string.csv](https://github.com/kuzudb/kuzu/files/13030485/long-string.csv)

### PR DESCRIPTION
**Pull request by acquamarin**
Closes #2232

### CODE REVIEW

**Comment by Riolku** [comment]
Rerunning gcc build because coverage uploading failed.

File: src/include/storage/store/var_list_column_chunk.h [hunk 1]
```diff
@@ -67,6 +66,9 @@ class VarListColumnChunk : public ColumnChunk {
     inline common::offset_t getListOffset(common::offset_t offset) const {
         return offset == 0 ? 0 : getValue<uint64_t>(offset - 1);
     }
+
+private:
```
Comment (Riolku) [inline src/include/storage/store/var_list_column_chunk.h:70]: Redundant access specifier.

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 2]
```diff
@@ -663,6 +663,10 @@ void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view
         if (!utf8proc::Utf8Proc::isValid(strVal.data(), strVal.length())) {
             throw common::CopyException{"Invalid UTF8-encoded string."};
         }
+        if (strVal.length() > BufferPoolConstants::PAGE_4KB_SIZE) {
+            throw CopyException(
+                ExceptionMessage::overLargeStringValueException(std::to_string(strVal.length())));
```
Comment (Riolku) [inline src/processor/operator/persistent/reader/csv/driver.cpp:668]: Let's refactor this function to take a `size_t` or `uint64_t`.

Comment (Riolku) [reply src/processor/operator/persistent/reader/csv/driver.cpp:668]: And on that note, is this the only call site? If so, inline it and delete the function.

File: src/storage/store/var_list_column_chunk.cpp [hunk 3]
```diff
@@ -86,14 +86,25 @@ void VarListColumnChunk::append(common::ValueVector* vector, common::offset_t st
...
+    auto numListValuesCopied = 0;
+    while (numListValuesCopied < numListValuesToCopy) {
+        auto numListValuesToCopyInBatch =
+            std::min<uint64_t>(numListValuesToCopy - numListValuesCopied, DEFAULT_VECTOR_CAPACITY);
+        dataVector->state->selVector->selectedSize = numListValuesToCopyInBatch;
+        for (auto j = 0u; j < numListValuesToCopyInBatch; j++) {
+            dataVector->state->selVector->selectedPositions[j] =
+                entry.offset + numListValuesCopied + j;
         }
         varListDataColumnChunk.append(dataVector);
```
Comment (Riolku) [inline src/storage/store/var_list_column_chunk.cpp:105]: Why does this code work, reusing the same vector?

File: test/test_files/copy/copy_edge_case.test [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
...760,9761,9762,9763,9764,9765,9766,9767,9768,9769,9770,9771,9772,9773,9774,9775,9776,9777,9778,9779,9780,9781,9782,9783,9784,9785,9786,9787,9788,9789,9790,9791,9792,9793,9794,9795,9796,9797,9798,9799,9800,9801,9802,9803,9804,9805,9806,9807,9808,9809,9810,9811,9812,9813,9814,9815,9816,9817,9818,9819,9820,9821,9822,9823,9824,9825,9826,9827,9828,9829,9830,9831,9832,9833,9834,9835,9836,9837,9838,9839,9840,9841,9842,9843,9844,9845,9846,9847,9848,9849,9850,9851,9852,9853,9854,9855,9856,9857,9858,9859,9860,9861,9862,9863,9864,9865,9866,9867,9868,9869,9870,9871,9872,9873,9874,9875,9876,9877,9878,9879,9880,9881,9882,9883,9884,9885,9886,9887,9888,9889,9890,9891,9892,9893,9894,9895,9896,9897,9898,9899,9900,9901,9902,9903,9904,9905,9906,9907,9908,9909,9910,9911,9912,9913,9914,9915,9916,9917,9918,9919,9920,9921,9922,9923,9924,9925,9926,9927,9928,9929,9930,9931,9932,9933,9934,9935,9936,9937,9938,9939,9940,9941,9942,9943,9944,9945,9946,9947,9948,9949,9950,9951,9952,9953,9954,9955,9956,9957,9958,9959,9960,9961,9962,9963,9964,9965,9966,9967,9968,9969,9970,9971,9972,9973,9974,9975,9976,9977,9978,9979,9980,9981,9982,9983,9984,9985,9986,9987,9988,9989,9990,9991,9992,9993,9994,9995,9996,9997,9998,9999]}
```
Comment (Riolku) [inline test/test_files/copy/copy_edge_case.test:13]: Don't we have a limit on the number of bytes in a list? How come this doesn't hit that limit?

File: test/test_files/copy/copy_edge_case.test [hunk 5]
```diff
@@ -0,0 +1,22 @@
+-GROUP CopyEdgeCaseTest
```
Comment (Riolku) [inline test/test_files/copy/copy_edge_case.test:1]: I don't think this file/group name makes sense. These aren't edge cases. These are long value tests.

File: test/test_files/copy/copy_edge_case.test [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
...99709719729739749759769779789799809819829839849859869879889899909919929939949959969979989991000100110021003100410051006100710081009101010111012101310141015101610171018101910201021102210231024102510261027102810291030103110321033103410351036103710381039104010411042104310441045104610471048104910501051105210531054105510561057105810591060106110621063106410651066106710681069107010711072107310741075107610771078107910801081108210831084108510861087108810891090109110921093109410951096109710981099110011011102110311041105110611071108110911101111111211131114111511161117111811191120112111221123112411251126112711281129113011311132113311341135113611371138113911401141114211431144114511461147114811491150115111521153115411551156115711581159116011611162116311641165116611671168116911701171117211731174117511761177117811791180118111821183118411851186118711881189119011911192119311941195119611971198119912001201120212031204120512061207120812091210121112121213121412151216121712181219122012211222122312241225122612271228122912301231123212331234123512361237123812391240124112421243124412451246124712481249125012511252125312541255125612571258125912601261126212631264126512661267126812691270127112721273127412751276}
```
Comment (Riolku) [inline test/test_files/copy/copy_edge_case.test:22]: Why test a string that's long but not long enough to trigger the bug I filed?

File: test/test_files/copy/copy_long_string.test [hunk 7]
```diff
@@ -1,20 +1,13 @@
...
----- 2
-${AliceStr}A
-Bob
-
--LOG Gender
--STATEMENT MATCH (p:person) RETURN p.gender
----- 2
-1
-2
+-LOG LongStringError
```
Comment (Riolku) [inline test/test_files/copy/copy_long_string.test:8]: Why is this file in a different place?

File: src/processor/operator/persistent/reader/csv/driver.cpp (hunk 2 above)
Comment (Riolku) [reply src/processor/operator/persistent/reader/csv/driver.cpp:668]: Should've checked.. It's not the only call site

File: src/include/storage/store/var_list_column_chunk.h (hunk 1 above)
Comment (acquamarin) [reply src/include/storage/store/var_list_column_chunk.h:70]: We always put an access specifier for functions and fields.

File: src/storage/store/var_list_column_chunk.cpp (hunk 3 above)
Comment (acquamarin) [reply src/storage/store/var_list_column_chunk.cpp:105]: The selectionVector can only store the state for 2048 values, we should only append 2048 values at a time when the list is large.

File: test/test_files/copy/copy_edge_case.test (hunk 4 above)
Comment (acquamarin) [reply test/test_files/copy/copy_edge_case.test:13]: We don't have a limit on list size.
This list doesn't hit the limit.

File: test/test_files/copy/copy_edge_case.test (hunk 6 above)
Comment (acquamarin) [reply test/test_files/copy/copy_edge_case.test:22]: See the test/test_files/copy/copy_long_string.test
The bug you reported is because someone removes the string length constraint. All strings must be smaller than 4KB

File: test/test_files/copy/copy_long_string.test (hunk 7 above)
Comment (acquamarin) [reply test/test_files/copy/copy_long_string.test:8]: This test should throw an error, they are different tests.

File: test/test_files/copy/copy_edge_case.test (hunk 5 above)
Comment (acquamarin) [reply test/test_files/copy/copy_edge_case.test:1]: We are testing when the strings and lists are long, i think this is an edge case.

File: src/storage/store/var_list_column_chunk.cpp (hunk 3 above)
Comment (Riolku) [reply src/storage/store/var_list_column_chunk.cpp:105]: Yeah. My actual confusion was I didn't realize `varlistdatacolumnChunk.append()` copies the valuevector

File: test/test_files/copy/copy_edge_case.test (hunk 4 above)
Comment (Riolku) [reply test/test_files/copy/copy_edge_case.test:13]: What about this? https://github.com/kuzudb/kuzu/blob/master/src/storage/storage_structure/disk_overflow_file.cpp#L239

Comment (acquamarin) [reply test/test_files/copy/copy_edge_case.test:13]: Lists are stored differently in node/rels.
Node: offset + data. There is no limit on the list size.
Rel: ku_list_t, maximum list size is 4K, but rel storage is currently changing by @ray6080 . After his PR, lists on rel should also have no size limitations.

File: src/processor/operator/persistent/reader/csv/driver.cpp (hunk 2 above)
Comment (acquamarin) [reply src/processor/operator/persistent/reader/csv/driver.cpp:668]: Added a check function.

File: src/storage/store/table_copy_utils.cpp [hunk 8]
```diff
@@ -507,6 +508,13 @@ std::unique_ptr<arrow::PrimitiveArray> TableCopyUtils::createArrowPrimitiveArray
     return std::make_unique<arrow::PrimitiveArray>(type, length, buffer);
 }
 
+void TableCopyUtils::validateStrLen(uint64_t strLen) {
+    if (strLen > BufferPoolConstants::PAGE_4KB_SIZE) {
+        throw CopyException(
+            ExceptionMessage::overLargeStringValueException(std::to_string(strLen)));
```
Comment (Riolku) [inline src/storage/store/table_copy_utils.cpp:514]: At this point just inline the function then.

Comment (acquamarin) [reply src/storage/store/table_copy_utils.cpp:514]: We don't inline any function that has more than one line

Comment (Riolku) [reply src/storage/store/table_copy_utils.cpp:514]: That's not what I mean. I mean, delete `overLargeStringValueException`, and here, do:

```cpp
throw CopyException(stringFormat("Maximum length of strings is 4096. Input string's length is {}.", strLen))
```

File: test/test_files/copy/copy_edge_case.test [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+-DATASET CSV empty
+
+--
+
+-CASE CopyLongListTest
+-STATEMENT CREATE NODE TABLE dog (id int64, counter int64[], primary key(id))
+---- ok
+-STATEMENT copy dog from "${KUZU_ROOT_DIRECTORY}/dataset/csv-edge-case-tests/long-list.csv"
+---- ok
+-STATEMENT match (d:dog) return *;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply test/test_files/copy/copy_edge_case.test:11]: I don't think it matters, why should we do this change?

**Comment by Riolku** [comment]
General comment, we should cancel CI runs that are redundant. I cancelled run #6875 for you.

File: test/test_files/copy/copy_edge_case.test (hunk 9 above)
[the remark this answers is not in the document]
Comment (Riolku) [reply test/test_files/copy/copy_edge_case.test:11]: If we change internal id format this test doesn't break. No reason to test more than we need to, cleaner test overall.

File: src/storage/store/table_copy_utils.cpp (hunk 8 above)
Comment (acquamarin) [reply src/storage/store/table_copy_utils.cpp:514]: We not include the exception message in the check function?

File: src/storage/store/table_copy_utils.cpp [hunk 10]
```diff
@@ -507,6 +508,13 @@ std::unique_ptr<arrow::PrimitiveArray> TableCopyUtils::createArrowPrimitiveArray
...
+    }
+}
+
```
Comment (Riolku) [inline src/storage/store/table_copy_utils.cpp:517]:
```suggestion
void TableCopyUtils::validateStrLen(uint64_t strLen) {
    if (strLen > BufferPoolConstants::PAGE_4KB_SIZE) {
        throw CopyException(stringFormat("Maximum length of strings is 4096. Input string's length is {}.", strLen))
    }
}
```

File: src/storage/store/table_copy_utils.cpp (hunk 8 above)
Comment (Riolku) [reply src/storage/store/table_copy_utils.cpp:514]: Sorry I don't know what you mean by that... I've posted a suggested change that might clear up confusion.

File: src/storage/store/table_copy_utils.cpp (hunk 10 above)
Comment (acquamarin) [reply src/storage/store/table_copy_utils.cpp:517]: I see, but that exception message is also used by update function, when doing string updates

Comment (Riolku) [reply src/storage/store/table_copy_utils.cpp:517]: I'd suggest refactoring `update()` to use `validateStrLen`.

Comment (acquamarin) [reply src/storage/store/table_copy_utils.cpp:517]: No, storage side shouldn't use this function. This function throws copyException while the storage should throw a runTimeException

Comment (Riolku) [reply src/storage/store/table_copy_utils.cpp:517]: Then just change the argument to an integer instead of a string and call it.

File: src/storage/store/string_column_chunk.cpp [hunk 11]
```diff
@@ -108,10 +108,7 @@ void StringColumnChunk::write(const Value& val, uint64_t posToWrite) {
 }
 
 void StringColumnChunk::setValueFromString(const char* value, uint64_t length, uint64_t pos) {
-    if (length > BufferPoolConstants::PAGE_4KB_SIZE) {
-        throw CopyException(
-            ExceptionMessage::overLargeStringValueException(std::to_string(length)));
-    }
+    TableCopyUtils::validateStrLen(length);
```
Comment (Riolku) [inline src/storage/store/string_column_chunk.cpp:111]: Let's add a test for this function.

File: src/storage/store/table_copy_utils.cpp (hunk 10 above)
Comment (acquamarin) [reply src/storage/store/table_copy_utils.cpp:517]: I don't understand what's the different between passing an int and a string?
If we pass an int, we also need to convert it to string inside the overLargeStringValueException.

File: src/storage/store/string_column_chunk.cpp (hunk 11 above)
Comment (acquamarin) [reply src/storage/store/string_column_chunk.cpp:111]: This function is only used for copyrel, and will be removed by @ray6080 quickly once the storage layout is changed.

**Review by Riolku — APPROVED** [review]
